In [96]:
import graphlab

In [97]:
people = graphlab.SFrame('people_wiki.gl/')

In [3]:
people.head()

Columns:
	URI	str
	name	str
	text	str

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+
|              text             |
+-------------------------------+
| digby morrell born 10 octo... |
| alfred j lewy aka sandy le... |
| harpdog brown is a singer ... |
| franz rottensteiner born i... |
| henry krvits born 30 decem... |
| sam henderson born october... |
| aaron lacrate is an americ... |
| trevor ferguson aka john f... |
| grant nelson born 27 april... |
| cathy caruth born 1955 is ... |
+-------------------------------+
[10 rows x 3 columns]

In [98]:
obama = people[people['name'] == 'Barack Obama']

In [5]:
obama

Columns:
	URI	str
	name	str
	text	str

Rows: Unknown

Data:
+-------------------------------+--------------+-------------------------------+
|              URI              |     name     |              text             |
+-------------------------------+--------------+-------------------------------+
| <http://dbpedia.org/resour... | Barack Obama | barack hussein obama ii br... |
+-------------------------------+--------------+-------------------------------+
[? rows x 3 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use len(sf) to force materialization.

In [99]:
obama['word_count'] = graphlab.text_analytics.count_words(obama['text'])

In [11]:
type(obama['word_count'])

graphlab.data_structures.sarray.SArray

In [10]:
type(obama[['word_count']])

graphlab.data_structures.sframe.SFrame

In [102]:
obama[['word_count','name']]

Columns:
	word_count	dict
	name	str

Rows: 1

Data:
+-------------------------------+--------------+
|           word_count          |     name     |
+-------------------------------+--------------+
| {'operations': 1, 'represe... | Barack Obama |
+-------------------------------+--------------+
[1 rows x 2 columns]

In [12]:
obama_word_count_table = obama[['word_count']].stack('word_count', new_column_name=['word','count'])

In [13]:
type(obama_word_count_table)

graphlab.data_structures.sframe.SFrame

In [14]:
obama_word_count_table

Columns:
	word	str
	count	int

Rows: 273

Data:
+------------+-------+
|    word    | count |
+------------+-------+
|    cuba    |   1   |
| relations  |   1   |
|   sought   |   1   |
|   combat   |   1   |
|   ending   |   1   |
| withdrawal |   1   |
|   state    |   1   |
|  islamic   |   1   |
|     by     |   1   |
|   gains    |   1   |
+------------+-------+
[273 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [15]:
obama_word_count_table.sort('count', ascending=False)

Columns:
	word	str
	count	int

Rows: 273

Data:
+-------+-------+
|  word | count |
+-------+-------+
|  the  |   40  |
|   in  |   30  |
|  and  |   21  |
|   of  |   18  |
|   to  |   14  |
|  his  |   11  |
| obama |   9   |
|  act  |   8   |
|   he  |   7   |
|   a   |   7   |
+-------+-------+
[273 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [25]:
test_sf = graphlab.SFrame({'f1':[{'a':[1,1],'b':[2,2]},{'a':1,'b':2},{'a':1,'b':2},{'a':1,'b':2},{'a':1,'b':2}],'f2':['pink','red','red','yellow','blue'],'label':[1,1,1,0,0]})

In [45]:
test_sf = graphlab.SFrame({'f1':[{'a':[1,1],'b':[2,2]}],'label':[1]})

In [48]:
test_sf[['f1']].stack('f1',new_column_name=['key','value'])

Columns:
	key	str
	value	array

Rows: 2

Data:
+-----+------------+
| key |   value    |
+-----+------------+
|  a  | [1.0, 1.0] |
|  b  | [2.0, 2.0] |
+-----+------------+
[2 rows x 2 columns]

In [49]:
people['word_count'] = graphlab.text_analytics.count_words(people['text'])

In [52]:
tfidf = graphlab.text_analytics.tf_idf(people['word_count'])
tfidf

Columns:
	docs	dict

Rows: 59071

Data:
+-------------------------------+
|              docs             |
+-------------------------------+
| {'selection': 3.8365785530... |
| {'precise': 6.443200606955... |
| {'just': 2.700729968710864... |
| {'all': 1.6431112434912472... |
| {'they': 1.899340117819389... |
| {'currently': 1.6370889691... |
| {'exclusive': 10.455187230... |
| {'taxi': 6.052021456094502... |
| {'houston': 3.935505942157... |
| {'phenomenon': 5.750053426... |
+-------------------------------+
[59071 rows x 1 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [53]:
people['tfidf'] = tfidf['docs']

In [58]:
obama = people[people['name']=='Barack Obama']
obama[['tfidf']].stack('tfidf',new_column_name=['word','tfidf']).sort('tfidf',ascending=False)

Columns:
	word	str
	tfidf	float

Rows: 273

Data:
+-------------+---------------+
|     word    |     tfidf     |
+-------------+---------------+
|    obama    | 43.2956530721 |
|     act     |  27.678222623 |
|     iraq    |  17.747378588 |
|   control   | 14.8870608452 |
|     law     | 14.7229357618 |
|   ordered   | 14.5333739509 |
|   military  | 13.1159327785 |
| involvement | 12.7843852412 |
|   response  | 12.7843852412 |
|  democratic | 12.4106886973 |
+-------------+---------------+
[273 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [59]:
clinton = people[people['name']=='Bill Clinton']

In [60]:
beckham = people[people['name']=='David Beckham']

In [66]:
knn_model = graphlab.nearest_neighbors.create(people,features=['tfidf'],label='name')

PROGRESS: Starting brute force nearest neighbors model training.


In [70]:
people[people['name']=='Barack Obama']

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: Unknown

Data:
+-------------------------------+--------------+-------------------------------+
|              URI              |     name     |              text             |
+-------------------------------+--------------+-------------------------------+
| <http://dbpedia.org/resour... | Barack Obama | barack hussein obama ii br... |
+-------------------------------+--------------+-------------------------------+
+-------------------------------+-------------------------------+
|           word_count          |             tfidf             |
+-------------------------------+-------------------------------+
| {'operations': 1, 'represe... | {'operations': 3.811771079... |
+-------------------------------+-------------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use len(sf) to force materialization.

## Question 1: Compare top words according to word counts to TF-IDF

In [71]:
john = people[people['name']=='Elton John']

In [78]:
john_word_count_table = john[['word_count']].stack('word_count',new_column_name=['word','count'])

In [79]:
john_word_count_table.sort('count', ascending=False)

Columns:
	word	str
	count	int

Rows: 255

Data:
+-------+-------+
|  word | count |
+-------+-------+
|  the  |   27  |
|   in  |   18  |
|  and  |   15  |
|   of  |   13  |
|   a   |   10  |
|  has  |   9   |
|  john |   7   |
|   he  |   7   |
|   on  |   6   |
| award |   5   |
+-------+-------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [80]:
john[['tfidf']].stack('tfidf',new_column_name=['word','tfidf']).sort('tfidf',ascending=False)

Columns:
	word	str
	tfidf	float

Rows: 255

Data:
+---------------+---------------+
|      word     |     tfidf     |
+---------------+---------------+
|    furnish    |  18.38947184  |
|     elton     |  17.48232027  |
|   billboard   | 17.3036809575 |
|      john     | 13.9393127924 |
|  songwriters  |  11.250406447 |
| tonightcandle | 10.9864953892 |
|  overallelton | 10.9864953892 |
|    19702000   | 10.2933482087 |
|   fivedecade  | 10.2933482087 |
|      aids     |  10.262846934 |
+---------------+---------------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

## Answer 1: 
* word_count: the, in, and
* tf-idf: furnish, elton, billboard

## Question 2: Measuring distance

In [81]:
beckham = people[people['name']=='Victoria Beckham']

In [83]:
graphlab.distances.cosine(john['tfidf'][0],beckham['tfidf'][0])

0.9567006376655429

In [84]:
paul = people[people['name']=='Paul McCartney']

In [86]:
graphlab.distances.cosine(john['tfidf'][0],paul['tfidf'][0])

0.8250310029221779

## Answer 2:
* The cosine distance between the articles on ‘Elton John’ and ‘Victoria Beckham’ is 0.9567006376655429.
* The cosine distance between 'Elton John' and 'Paul McCartney' is 0.8250310029221779.

## Question 3: Building the nearest neighbors models with different input features and setting the distance metric

In [90]:
tfidf_nn_model = graphlab.nearest_neighbors.create(people,features=['tfidf'],distance='cosine',label='name')

PROGRESS: Starting brute force nearest neighbors model training.


In [91]:
wc_nn_model = graphlab.nearest_neighbors.create(people,features=['word_count'],distance='cosine',label='name')

PROGRESS: Starting brute force nearest neighbors model training.


In [92]:
wc_nn_model.query(john)

PROGRESS: Starting pairwise querying.
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | Query points | # Pairs | % Complete. | Elapsed Time |
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | 0            | 1       | 0.00169288  | 47.79ms      |
PROGRESS: | Done         |         | 100         | 574.129ms    |
PROGRESS: +--------------+---------+-------------+--------------+


Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-------------------+-------------------+------+
| query_label |  reference_label  |      distance     | rank |
+-------------+-------------------+-------------------+------+
|      0      |     Elton John    | 2.22044604925e-16 |  1   |
|      0      |   Cliff Richard   |   0.16142415259   |  2   |
|      0      |   Sandro Petrone  |   0.16822542751   |  3   |
|      0      |    Rod Stewart    |   0.168327165587  |  4   |
|      0      | Malachi O'Doherty |   0.177315545979  |  5   |
+-------------+-------------------+-------------------+------+
[5 rows x 4 columns]

In [93]:
tfidf_nn_model.query(john)

PROGRESS: Starting pairwise querying.
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | Query points | # Pairs | % Complete. | Elapsed Time |
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | 0            | 1       | 0.00169288  | 90ms         |
PROGRESS: | Done         |         | 100         | 1.00s        |
PROGRESS: +--------------+---------+-------------+--------------+


Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+------------------+--------------------+------+
| query_label | reference_label  |      distance      | rank |
+-------------+------------------+--------------------+------+
|      0      |    Elton John    | -2.22044604925e-16 |  1   |
|      0      |   Rod Stewart    |   0.717219667893   |  2   |
|      0      |  George Michael  |   0.747600998969   |  3   |
|      0      | Sting (musician) |   0.747671954431   |  4   |
|      0      |   Phil Collins   |   0.75119324879    |  5   |
+-------------+------------------+--------------------+------+
[5 rows x 4 columns]

In [94]:
wc_nn_model.query(beckham)

PROGRESS: Starting pairwise querying.
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | Query points | # Pairs | % Complete. | Elapsed Time |
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | 0            | 1       | 0.00169288  | 38.154ms     |
PROGRESS: | Done         |         | 100         | 662.27ms     |
PROGRESS: +--------------+---------+-------------+--------------+


Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+--------------------------+--------------------+------+
| query_label |     reference_label      |      distance      | rank |
+-------------+--------------------------+--------------------+------+
|      0      |     Victoria Beckham     | -2.22044604925e-16 |  1   |
|      0      | Mary Fitzgerald (artist) |   0.207307036115   |  2   |
|      0      |      Adrienne Corri      |   0.214509782788   |  3   |
|      0      |     Beverly Jane Fry     |   0.217466468741   |  4   |
|      0      |      Raman Mundair       |   0.217695474992   |  5   |
+-------------+--------------------------+--------------------+------+
[5 rows x 4 columns]

In [95]:
tfidf_nn_model.query(beckham)

PROGRESS: Starting pairwise querying.
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | Query points | # Pairs | % Complete. | Elapsed Time |
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | 0            | 1       | 0.00169288  | 91.708ms     |
PROGRESS: | Done         |         | 100         | 934.274ms    |
PROGRESS: +--------------+---------+-------------+--------------+


Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+---------------------+-------------------+------+
| query_label |   reference_label   |      distance     | rank |
+-------------+---------------------+-------------------+------+
|      0      |   Victoria Beckham  | 1.11022302463e-16 |  1   |
|      0      |    David Beckham    |   0.548169610263  |  2   |
|      0      | Stephen Dow Beckham |   0.784986706828  |  3   |
|      0      |        Mel B        |   0.809585523409  |  4   |
|      0      |    Caroline Rush    |   0.819826422919  |  5   |
+-------------+---------------------+-------------------+------+
[5 rows x 4 columns]

## Answer 3:
* John - word count cosine distance : Cliff Richard
       - tfidf cosine distance : Rod Stewart
* beckham - word count cosine distance : Mary Fitzgerald (artist)
          - tfidf cosine distance : David Beckham